mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Zne.hpp"
+
+#include <algorithm>
+#include <deque>
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include "llvm/ADT/SmallPtrSet.h"
+
+#include "Mitigation/IR/MitigationOps.h"
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Utils/RemoveQuantum.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/IRMapping.h"
+
+namespace catalyst {
+namespace mitigation {
+
+LogicalResult ZneLowering::match(mitigation::ZneOp op) const { return success(); }
+
+void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
+{
+    Location loc = op.getLoc();
+
+    // Scalar factors
+    auto scalarFactors = op.getScalarFactors();
+    RankedTensorType scalarFactorType = scalarFactors.getType().cast<RankedTensorType>();
+    auto sizeInt = scalarFactorType.getDimSize(0);
+
+    // Create the folded circuit function
+    FlatSymbolRefAttr foldedCircuitRefAttr =
+        getOrInsertFoldedCircuit(loc, rewriter, op, scalarFactorType.getElementType());
+    func::FuncOp foldedCircuit =
+        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, foldedCircuitRefAttr);
+    RankedTensorType resultType = op.getResultTypes().front().cast<RankedTensorType>();
+
+    // Loop over the scalars to create a folded circuit per factor
+    Value c0 = rewriter.create<index::ConstantOp>(loc, 0);
+    Value c1 = rewriter.create<index::ConstantOp>(loc, 1);
+    Value size = rewriter.create<index::ConstantOp>(loc, sizeInt);
+    // Initialize the results as empty tensor
+    Value results =
+        rewriter.create<tensor::EmptyOp>(loc, resultType.getShape(), resultType.getElementType());
+    Value resultValues =
+        rewriter
+            .create<scf::ForOp>(
+                loc, c0, size, c1, /*iterArgsInit=*/results,
+                [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+                    std::vector<Value> newArgs{op.getArgs().begin(), op.getArgs().end()};
+                    SmallVector<Value> index = {i};
```

Comment:
What's the requirement of this `SmallVector`? Couldn't we use `{i}` in lines 70 and 76 instead?

Reply:
It needs to be a range in extractop, therefore I used small vector. Using value range has brought some trouble with lifetime scopes in the past.

Reply:
thanks!